GoogleForCreators/web-stories-wp
# Design System Storybook: Tooltips

## Summary

Tooltip component based on figma added to storybook.

## Relevant Technical Choices
- ~Takes dashboard tooltip and adds extra positions to it like can be found in existing editor tooltip~
- ~Adds functionality to show tooltip on relevant focus (like editor)~ 
- Uses a clip-path for tooltip tail because it's got curves 
(Update 19/11/2020)
- After a few edge cases were pointed out, decided that even though it adds complexity it'll make swapping in the updated tooltips a whole lot easier on the editor side of things, to swap implementation and follow the structure set up in the editor closely. Reasons being: scrolling and monitoring space until browser edge. The tooltips in the editor have already been robustly tested. 
- I have copied in some resources from the edit-story directory to the design-system to use the tooltip structure there. These include: 
    - keyboard utils for `isPlatformMacOS` and `prettifyShortcut` 
    - The `popup` component which I copied verbatim with 2 exceptions: 1) I commented out the style overrides for scrollbars since those will need to be adjusted with new designs separately - left a note 2) organized the popup utils and constants. 
    - `utils/useResizeEffect` which we use in both edit-story and dashboard - the popup needs this util. 
- From here, I brought in the updated styles and tail as a clippath - I moved the tail to its own file to help separate concerns and maintain readability and exported the popup positions as `TOOLTIP_POSITIONS` to follow same structure as buttons in design system so that when using tooltips in a codebase the constants are found central to it without having to know that popup is involved. 

## To-do

- [x] [Confirming proper background color](https://github.com/google/web-stories-wp/issues/4958#issuecomment-729227557) with Sam
- More robust tests would be nice with Karma eventually. Testing positions with RTL is tricky.

## User-facing changes

None, storybook only

## Testing Instructions

Go to storybook > designSystem > components > tooltip and see both dark and light mode tooltips. They should mirror [figma](https://www.figma.com/file/bMhG3KyrJF8vIAODgmbeqT/Design-System?node-id=3398%3A83619).


Demo with keyboard focus
![lightmode keyboard demo](https://user-images.githubusercontent.com/10720454/99714488-e0b43480-2a62-11eb-9d86-88b5d250dbcc.gif)

Demo with cursor
![lightmode demo](https://user-images.githubusercontent.com/10720454/99714522-e9a50600-2a62-11eb-9ace-58f273798eec.gif)
![darkmode demo](https://user-images.githubusercontent.com/10720454/99714527-ec9ff680-2a62-11eb-8459-17340d85609b.gif)


---



Addresses #4958
